power_grid_model_c/power_grid_model/include/power_grid_model/sparse_mapping.hpp

```diff
@@ -50,35 +50,109 @@ struct SparseMapping {
     // copy back
+    // copy everything except the indptr[0] to counter.
+    // counter becomes {0, 2, 4, 5, 5, 6, 6}
     std::copy(sparse_mapping.indptr.cbegin() + 1, sparse_mapping.indptr.cend(), counter.begin());
     // sort
+    // crbegin() points to the element at the end of the container
+    // crend() points to the position before the first element
+    // at this point: entries[{0,3},{1,5},{2,2},{3,1},{4,1},{5,2}], counter {0, 2, 4, 5, 5, 6, 6}
+    // Reverse order the first entry in entries: {5,2}, it_entry->second = 2
+    // counter[2] is 4, decremented to 3(counter becomes {0, 2, 3, 5, 5, 6, 6}), thus sparse_mapping.reorder[3] is set
+    // to it_entry->first which is 5. sparse_mapping.reorder = {0, 0, 0, 5, 0, 0},
+    // Final result: sparse_mapping.reorder = {3, 4, 2, 5, 0, 1}
     for (auto it_entry = entries.crbegin(); it_entry != entries.crend(); ++it_entry) {
         sparse_mapping.reorder[--counter[it_entry->second]] = it_entry->first;
     }
+    // the first is the index before reordering of load_gen
+    // it entry second is the index of the bus
+    // reorder at the end contains for each new index of the load gen the old index of the load gen in a linear way
     assert(sparse_mapping.indptr[0] == 0);
     assert(sparse_mapping.indptr.back() == n_A);
     return sparse_mapping;
 }
 
-} // namespace power_grid_model
+struct DenseMapping {
+    IdxVector indvector;
+    IdxVector reorder;
+};
+
+inline DenseMapping build_dense_mapping(IdxVector const& idx_B_in_A, Idx const n_B) {
+    auto const n_A = static_cast<Idx>(idx_B_in_A.size());
+
+    using DenseEntry = std::pair<Idx, Idx>;
+    // entries vector with size n_A
+    std::vector<DenseEntry> entries(n_A);
+    // idx_B_in_A{3, 5, 2, 1, 1, 2}, then count = {0, 1, 2, 3, 4, 5} since count is affected by the length of
+    // idx_B_in_A j_B is an element from idx_B_in_A, and i_A is the current value from IdxCount{0} returns
+    // entries[{0,3},{1,5},{2,2},{3,1},{4,1},{5,2}]
+    std::transform(idx_B_in_A.cbegin(), idx_B_in_A.cend(), IdxCount{0}, entries.begin(), [](Idx j_B, Idx i_A) {
+        return DenseEntry{i_A, j_B};
+    });
+
+    DenseMapping dense_mapping;
+
+    dense_mapping.indvector.resize(n_A);
+    // size of reorder will be 6
+    dense_mapping.reorder.resize(n_A);
+
+    IdxVector xndvector;
+    xndvector.resize(n_B);
+
+    IdxVector counter(n_B, 0);
 
+    for (auto const& entry : entries) {
+        ++counter[entry.second];
+    }
+
+    std::inclusive_scan(counter.cbegin(), counter.cend(), xndvector.begin());
+
+    std::copy(xndvector.cbegin(), xndvector.cend(), counter.begin());
+
+    for (auto i = n_A - 1; i >= 0; i--) {
+        dense_mapping.indvector[xndvector[idx_B_in_A[i]] - 1] = idx_B_in_A[i];
+        xndvector[idx_B_in_A[i]]--;
+    }
+
+    for (auto it_entry = entries.crbegin(); it_entry != entries.crend(); ++it_entry) {
+        dense_mapping.reorder[--counter[it_entry->second]] = it_entry->first;
+    }
```

Comment:
there's also a different approach:

what we seek is the sorted list `[a_i]` of the input values `{a_i | i in [0, ..., n]}` together with the (reordered) indices of the original list `[0, 1, ..., n]`.

There are 2 different algorithms that may be efficient, depending on the size `n` and the highest entry `r = max{a_i | i in [0, ..., n]}` (i use the convention in https://en.wikipedia.org/wiki/Sorting_algorithm#Non-comparison_sorts ).

if `n` is of the order of or greater than `r`, then just using `build_sparse_mapping` and then using `sparse_decode` on the `indptrs` suffices. This is the [Counting Sort](https://en.wikipedia.org/wiki/Counting_sort) you implemented. To shorted the code, you can also just do

```cpp
auto const sparse_mapping = build_sparse_mapping(idx_B_in_A, n_B);
return DenseMapping {
    .values = sparse_decode(sparse_mapping.indptr),
    .reorder = std::move(sparse_mapping.reorder)
};
```

The additional computational overhead is of the order of `r` and not any worse than the existing implementation (because that's exactly what happens when you do `DenseGroupedIdxVector{from_sparse, sparse_mapping.indptr}`).

Otherwise, `n << r`. This means that if you `sort` a vector of pairs containing both the original index and the input value to sort based on the value, then you automatically obtain the correct ordering. E.g. (removing any reserves, moving, etc.; just for illustration)

```cpp
std::vector<std::pair<Idx, Idx>> mapping_to_from;
std::ranges::transform(idx_B_in_A,
                       std::vies::iota(0, static_cast<Idx>(idx_B_in_A.size()),
                       std::back_inserter(mapping_to_from),
                       [](Idx value, Idx orig_idx){ return std::pair{value, orig_idx}; });
std::ranges::sort(mapping_to_from);  // runs in O(n log n)

DenseMapping result;
std::ranges::transform(mapping_to_from, std::back_inserter(result.indvector),
                       [](std::pair<Idx, Idx> const& to_from) { return to_from.first; } );
std::ranges::transform(mapping_to_from, std::back_inserter(result.reorder),
                       [](std::pair<Idx, Idx> const& to_from) { return to_from.second; });
```

Now it just so happens that in most cases where we use `DenseGroupedIdxVector`, `n << r` (e.g.: there will usually be only one or a couple sources in a power grid). In that case, the first algorithm becomes `O(r)` and the second one remains `O(n log n)`. Since `n << r`, the second algorithm will be much faster.

I would recommend doing both implementations, keeping both of them as short as possible, with an `if` statement selecting one or the other based on whether `n log n < c * r` (but to do so, please calculate for each method the amount of operations, so that you have the correct threshold `c`)